### PR TITLE
Merge train: qualify the test branch by PR id

### PR DESCRIPTION
**Pull request by rudymatela**
Partly closes: #77 

Qualify the test branch by PR id in order to prepare for a future change where we will have multiple PR ids.  Instead of building on `testing`, we build on `testing/<id>`.

This is made on top of #131.

### Side effect

There is a little side effect of this, while the tests for a PR is running, GitHub will show a "compare & pull request" suggestion box:

compare-and-pull-request
compare-and-pull-request-w

It should disappear after the build passes or fails and the branch is deleted by Hoff.

Ideally, there should be a setting on GitHub to tell it to ignore branches starting with a given `testing/` prefix.  But it seems this is not the case.  According to my (lack of) findings, there is no way to set this:

* https://stackoverflow.com/questions/71162762/how-to-get-rid-of-githubs-compare-pull-request-suggestion
* https://github.community/t/how-to-fix-compare-pull-request/161488
* [https://duckduckgo.com/?q=github+dismiss+"compare+%26+pull+request"](https://duckduckgo.com/?q=github+dismiss+"compare+%26+pull+request")
* [https://www.google.com/search?q=github+dismiss+"compare+%26+pull+request"](https://www.google.com/search?q=github+dismiss+"compare+%26+pull+request")

One could think of mitigating this by having a small set of testing slots for which we start by dismissing manually:

* `testing/1`
* `testing/2`
* `testing/3`
* `testing/4`
* `testing/5`
* `testing/6`

However, this solution would not work, because:

* limit the maximum number of PRs in our merge train (#77);
* make it difficult to manage in terms of code (keeping track of free slots);
* make it difficult to know which testing branches are actually being used in case of build failures (with PR numbers it is easy!);
* ... all of this by not actually solving the problem from the get-go -- at least in the _first six builds per project_ we would have to manually dismiss each "compare and pull request" suggestion.

So it seems we are stuck with the minor side effect of seeing that button while the build builds.

### CODE REVIEW

File: src/Git.hs
Comment (rudymatela) [reply src/Git.hs:266]: Nevermind, I'm keeping it...

**Comment by ruuda** [comment]
> Qualify the test branch by PR id in order to prepare for a future change where we will have multiple PR ids

The PR id might not uniquely identify the thing you are building. Suppose we have PRs A and B, we build a train `master <- A <- B`, build `A` and `B`, but now `A` fails. Then we want to do `master <- B`, a different commit. I think if the strategy is to only build sequentially, and you don’t care about results of the pending builds on top of a failed build, then every pull request will occur at most once at the tip of a build, and it still works. But if you want to “hedge the bet” (e.g. build both `master <- A <- B`, and also `master <- B`, so that if `A` fails, then next best alternative is already in progress), a pull request can be at the tip of multiple builds.

> One could think of mitigating this by having a small set of testing slots

Yeah, I think you will need this either way. The more parallel builds you speculatively start, the lower the return on investment, because the longer the train is, the more likely it is to fail. And if you don’t limit parallelism in Hoff but instead let the CI provider handle that, then priority is hard to control. I would go with the build slots, and make the amount configurable.

> `testing/1`

Keep in mind that the branch `testing` and `testing/1` cannot coexist, because Git tracks branches as files under `.git/refs`, and `.git/refs/remotes/origin/testing` cannot simultaneously be a file and a directory. When the naming scheme changes, it will be a one-time nuisance for all people who pull from the repository and who might have a `testing` (remote tracking) branch locally, and the error can be puzzling if you’re not aware of this implementation detail leak.

**Comment by rudymatela** [comment]
> > Qualify the test branch by PR id in order to prepare for a future change where we will have multiple PR ids
> 
> The PR id might not uniquely identify the thing you are building. Suppose we have PRs A and B, we build a train `master <- A <- B`, build `A` and `B`, but now `A` fails. Then we want to do `master <- B`, a different commit. I think if the strategy is to only build sequentially, and you don’t care about results of the pending builds on top of a failed build, then every pull request will occur at most once at the tip of a build, and it still works. But if you want to “hedge the bet” (e.g. build both `master <- A <- B`, and also `master <- B`, so that if `A` fails, then next best alternative is already in progress), a pull request can be at the tip of multiple builds.

The [current idea](https://github.com/channable/hoff/issues/77) is to build only:

* master <- A
* master <- A <- B
* master <- A <- B <- C

The following will only start building once A fails (and consequently B and C, since the current plan does not allow merging further PRs on top of a failing PR):

* master <- B
* master <- B <- C

In this way the PR id still uniquely identifies the build.

In the future, if we decide to move towards more speculative builds I would rather use `testing/1-2-3` to signify three built on top of two built on top of one.

> Yeah, I think you will need this either way. The more parallel builds you speculatively start, the lower the return on investment, because the longer the train is, the more likely it is to fail. And if you don’t limit parallelism in Hoff but instead let the CI provider handle that, then priority is hard to control. I would go with the build slots, and make the amount configurable.

See the idea above: `testing/1-2-3`.

> Keep in mind that the branch `testing` and `testing/1` cannot coexist, because Git tracks branches as files under `.git/refs`, and `.git/refs/remotes/origin/testing` cannot simultaneously be a file and a directory. When the naming scheme changes, it will be a one-time nuisance for all people who pull from the repository and who might have a `testing` (remote tracking) branch locally, and the error can be puzzling if you’re not aware of this implementation detail leak.

I am aware of that, the current implementation already takes this into account, see: https://github.com/channable/hoff/pull/134/commits/503461e0cf048b62992aa3322b034fc25d12078a.  The `testing` branch is deleted before the creation of any `testing/<id>` branch.  That's also listed in the TODO list of #77:

* [x] fix issue of pushing `testing/<id>` in the presence of `testing`

**Comment by ruuda** [comment]
> The testing branch is deleted before the creation of any testing/<id> branch

Yes, but it will also be an issue for everybody who has cloned the repository and who has a remote-tracking branch (which is probably going to be everybody who uses that repository), it will make `git fetch` complain until the local ref is deleted:

```
$ git fetch
error: cannot lock ref 'refs/remotes/origin/testing/1': 'refs/remotes/origin/testing' exists; cannot create 'refs/remotes/origin/testing/1'
From /tmp/upstream
 ! [new branch]      testing/1  -> origin/testing/1  (unable to update local ref)
```

It’s nothing that an announcement can’t fix, I just wanted to point it out.

**Review by fatho — COMMENTED** [review]
LGTM for now - But I'll have a final look once it has been rebased after the other relevant PRs have been merged.

**Comment by rudymatela** [comment]
> > The testing branch is deleted before the creation of any testing/ branch
> 
> Yes, but it will also be an issue for everybody who has cloned the repository and who has a remote-tracking branch (which is probably going to be everybody who uses that repository), it will make `git fetch` complain until the local ref is deleted:
> 
> ```
> $ git fetch
> error: cannot lock ref 'refs/remotes/origin/testing/1': 'refs/remotes/origin/testing' exists; cannot create 'refs/remotes/origin/testing/1'
> From /tmp/upstream
>  ! [new branch]      testing/1  -> origin/testing/1  (unable to update local ref)
> ```
> 
> It’s nothing that an announcement can’t fix, I just wanted to point it out.

@ruuda This makes sense, noted.

I haven't tested this, but I guess this can be (manually) fixed in two ways:

1. `git fetch --prune`
2. `git branch -rd origin/testing && git fetch`

I'll check whether either/both work and include these in an announcement accordingly.

**Comment by rudymatela** [comment]
> LGTM for now - But I'll have a final look once it has been rebased after the other relevant PRs have been merged.

@fatho The other relevant PRs have been merged and I rebased on top of the new master.  Can you please take a look again?

**Review by fatho — APPROVED** [review]
LGTM

**Comment by rudymatela** [comment]
@fatho Thanks for the review.

I will wait [until I deploy v0.25.9](https://github.com/channable/devops/pull/5740) before I merge this.

**Comment by rudymatela** [comment]
I tested this in conjunction with GitHub again just to be sure and all seems to be working.  It's time to prepare the next deployment again.  @OpsBotPrime merge on Friday.

**Comment by OpsBotPrime** [comment]
Pull request approved for merge by @rudymatela, rebasing now.

**Comment by OpsBotPrime** [comment]
Rebased as 9cbd57e2bda4b03da7128183ec4ef027bc322a58, waiting for CI …

**Comment by OpsBotPrime** [comment]
Waiting on CI job: https://channable.semaphoreci.com/workflows/5c426238-cd7d-478a-8bd9-9c8bad8e6bbe?pipeline_id=b7fdb860-0ddc-4c33-9843-9cab068ea3a3

**Comment by rudymatela** [comment]
Here is the message I will send to the users.

Manual intervention required for @OpsBotPrime (Hoff v0.26.0)
------------------------------------

Hoff v0.26.0 builds PRs on separate testing branches (`testing/<pr-id>`).  After Hoff starts a testing build, you may get the following error when fetching or pulling from the remote repository:

```bash
$ git fetch
error: cannot lock ref 'refs/remotes/origin/testing/109': 'refs/remotes/origin/testing' exists; cannot create 'refs/remotes/origin/testing/109'
From github.com:organization/repository
 ! [new branch]      rudy-testing/109 -> origin/rudy-testing/109  (unable to update local ref)
error: some local refs could not be updated; try running
 'git remote prune origin' to remove any old, conflicting branches
```

If that happens, simply follow the instruction by git to prune the conflicting branch:

```bash
$ git remote prune origin
Pruning origin
URL: git@github.com:channable/git-sandbox
 * [pruned] origin/rudy-testing
```

Then pull or fetch again:

```bash
$ git fetch
```

Alternatively you can use `git fetch --prune`.

Unfortunately, this will happen once per repository.  To fix the issue for all repositories that have been built by Hoff already, you can do:

```bash
$ cd <workspace>
$ for repo in *; do git -C $repo fetch --prune; done
```

**Pull request merged**
